In [228]:
#import the necessary libraries
import os
import pandas as pd
import numpy as np
import json

In [229]:
#construct the requried dataframes for the two json files
pcd = pd.read_json("../HeroesOfPymoli/purchase_data.json")
pcd.head()

Age Gender  Item ID                                  Item Name  Price  \
0   38   Male      165                Bone Crushing Silver Skewer   3.37   
1   21   Male      119  Stormbringer, Dark Blade of Ending Misery   2.32   
2   34   Male      174                            Primitive Blade   2.46   
3   21   Male       92                               Final Critic   1.36   
4   23   Male       63                             Stormfury Mace   1.27   

             SN  
0     Aelalis34  
1        Eolo46  
2   Assastnya25  
3  Pheusrical25  
4        Aela59

In [230]:
#shape of the dataframes
pcd.shape

(780, 6)

In [231]:
#datatypes of the dataframe
pcd.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 780 entries, 0 to 779
Data columns (total 6 columns):
Age          780 non-null int64
Gender       780 non-null object
Item ID      780 non-null int64
Item Name    780 non-null object
Price        780 non-null float64
SN           780 non-null object
dtypes: float64(1), int64(2), object(3)
memory usage: 42.7+ KB


======================================================================================================================

**Player Count**

* Total Number of Players

In [241]:
# Total Number of Players
# pcd['SN'].nunique()

pcd.SN.value_counts().count()

573

In [260]:
pd.DataFrame({"Total Number of Players":pcd.SN.value_counts().count()},index=[0])

Total Number of Players
0                      573

======================================================================================================================

**Purchasing Analysis (Total)**

* Number of Unique Items
* Average Purchase Price
* Total Number of Purchases
* Total Revenue

In [261]:
# Number of Unique Items
pcd['Item ID'].nunique()

183

In [262]:
# Average Purchase Price
pcd['Price'].mean()

2.931192307692303

In [263]:
# Total Number of Purchases
len(pcd)

780

In [264]:
# Total Revenue
pcd['Price'].sum()

2286.3299999999963

In [265]:
pd.DataFrame({"Number of Unique Items":pcd['Item ID'].nunique(),
  "Average Purchase Price":pcd['Price'].mean(),
 "Total Number of Purchases":len(pcd),
"Total Revenue":pcd['Price'].sum()},index=[0])

Average Purchase Price  Number of Unique Items  Total Number of Purchases  \
0                2.931192                     183                        780   

   Total Revenue  
0        2286.33

======================================================================================================================

**Gender Demographics**

* Percentage and Count of Male Players
* Percentage and Count of Female Players
* Percentage and Count of Other / Non-Disclosed

In [269]:
# get all the duplicate rows
print(df.loc[df.duplicated("SN",keep=False),:].sort_values("SN").head(10))
print(len(df.loc[df.duplicated("SN",keep=False),:]))

     Age Gender  Item ID                             Item Name  Price  \
431   37   Male       92                          Final Critic   1.36   
308   37   Male       79                   Alpha, Oath of Zeal   2.88   
377   37   Male      174                       Primitive Blade   2.46   
647   26   Male      156          Soul-Forged Steel Shortsword   1.16   
721   26   Male       39  Betrayal, Whisper of Grieving Widows   2.35   
224   26   Male      106                   Crying Steel Sickle   2.29   
529   38   Male      172                    Blade of the Grave   1.69   
0     38   Male      165           Bone Crushing Silver Skewer   3.37   
359   20   Male       32                               Orenmir   4.95   
637   20   Male       18            Torchlight, Bond of Storms   1.77   

             SN  
431  Aduephos78  
308  Aduephos78  
377  Aduephos78  
647   Aeduera68  
721   Aeduera68  
224   Aeduera68  
529   Aelalis34  
0     Aelalis34  
359  Aeliriam77  
637  Aeliriam77 

In [274]:
# drop the duplicate SN rows to get gender propotions
pcd_unique = pcd.drop_duplicates("SN", keep='first')
len(pcd_unique)

573

In [273]:
pd.DataFrame({'Percentage of players' : pcd_unique['Gender'].value_counts(normalize=True),
   'Total Count' : pcd_unique['Gender'].value_counts()})

Percentage of players  Total Count
Male                                0.811518          465
Female                              0.174520          100
Other / Non-Disclosed               0.013962            8

======================================================================================================================

**Purchasing Analysis (Gender)** 

* The below each broken by gender
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
  * Normalized Totals (purchase total and divide it by the total count)

In [275]:
summary_gender_purchasing = pd.DataFrame(pcd.groupby('Gender').agg({'Price':[np.mean,np.sum],'Item ID':'count'}))
summary_gender_purchasing

Price          Item ID
                           mean      sum   count
Gender                                          
Female                 2.815515   382.91     136
Male                   2.950521  1867.68     633
Other / Non-Disclosed  3.249091    35.74      11

======================================================================================================================

**Age Demographics**

* The below each broken into bins of 4 years (i.e. &lt;10, 10-14, 15-19, etc.) 
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value
  * Normalized Totals

In [280]:
test = pd.DataFrame({'days': [0,31,45,60]})
test['range'] = pd.cut(test.days, [0,30,60,90], right=False)
test

# right : Indicates whether the bins include the rightmost edge or not. If
# right == True (the default), then the bins [1,2,3,4] indicate
# (1,2], (2,3], (3,4]

days     range
0     0   [0, 30)
1    31  [30, 60)
2    45  [30, 60)
3    60  [60, 90)

In [285]:
age_bins = [0, 10, 15, 20, 25, 30, 35, 40, 100]
age_labels = ['<10', '10-14', '15-19', '20-24','25-29','30-34','35-39','40+']
pcd['Age_Group'] = pd.cut(pcd['Age'], age_bins ,labels = age_labels,right=False)

In [286]:
summary_age = pd.DataFrame(pcd.groupby('Age_Group').agg({'Price':[np.mean,np.sum],'Item ID':'count'}))

In [287]:
summary_age

Price         Item ID
               mean     sum   count
Age_Group                          
10-14      2.770000   96.95      35
15-19      2.905414  386.42     133
20-24      2.913006  978.77     336
25-29      2.962640  370.33     125
30-34      3.082031  197.25      64
35-39      2.842857  119.40      42
40+        3.161765   53.75      17
<10        2.980714   83.46      28

======================================================================================================================

**Top Spenders**

* Identify the the top 5 spenders in the game by total purchase value, then list (in a table):
  * SN
  * Purchase Count
  * Average Purchase Price
  * Total Purchase Value

In [288]:
summary_spenders = pd.DataFrame(pcd.groupby(['SN']).agg({'Price':[np.mean,np.sum],'Item ID':'count'}))

In [289]:
summary_spenders.sort_values(('Price','sum'),ascending=False)[:5]

Price        Item ID
                 mean    sum   count
SN                                  
Undirrala66  3.412000  17.06       5
Saedue76     3.390000  13.56       4
Mindimnya67  3.185000  12.74       4
Haellysu29   4.243333  12.73       3
Eoda93       3.860000  11.58       3

======================================================================================================================

**Most Popular Items**

* Identify the 5 most popular items by purchase count, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value

In [290]:
summary_items = pd.DataFrame(pcd.groupby(['Item ID','Item Name']).agg({'Price':[np.mean,np.sum],'Item ID':'count'}))

In [291]:
summary_items.sort_values(('Item ID','count'),ascending=False)[0:5]

Price        Item ID
                                              mean    sum   count
Item ID Item Name                                                
39      Betrayal, Whisper of Grieving Widows  2.35  25.85      11
84      Arcane Gem                            2.23  24.53      11
31      Trickster                             2.07  18.63       9
175     Woeful Adamantite Claymore            1.24  11.16       9
13      Serenity                              1.49  13.41       9

======================================================================================================================

**Most Profitable Items**

* Identify the 5 most profitable items by total purchase value, then list (in a table):
  * Item ID
  * Item Name
  * Purchase Count
  * Item Price
  * Total Purchase Value

In [364]:
summary_items = pd.DataFrame(pcd.groupby(['Item ID','Item Name']).agg({'Price':[np.mean,np.sum],'Item ID':'count'}))

In [292]:
summary_items.sort_values(('Price','sum'),ascending=False)[0:5]

Price        Item ID
                                    mean    sum   count
Item ID Item Name                                      
34      Retribution Axe             4.14  37.26       9
115     Spectral Diamond Doomblade  4.25  29.75       7
32      Orenmir                     4.95  29.70       6
103     Singed Scalpel              4.87  29.22       6
107     Splitter, Foe Of Subtlety   3.61  28.88       8

======================================================================================================================

### Observations

    As per the Game's purchasing data, 81% of men made a purchase, compared to 17% of women indicating that the game is more popular among the male demographics. 

    Those consumers aged 20 to 24, remain the key age demographic for pymoli, spending more money than any other age group.
    
    Those consumers aged 40 and above remain the age group purchasing high value items.